### ISSUE TITLE
commit-msg: subject length is measured in bytes and reported as characters, and the body check changes unit with the locale

### ISSUE THREAD

**Issue by ruleofcodedev** [Rule-Of-Code/RoC#128]
**Version:** 7.21.0
**Component:** the generated `commit-msg` hook — the template in `GitHooksInstaller` at `dist/hooks/installer.js:224` and `:249`, emitted as `.husky/commit-msg:57` and `:82`

Kind: a wrong number printed by a gate. The subject limit is measured in **bytes** and reported as **characters**; the body limit is measured in whichever unit the ambient locale happens to pick, so the same commit passes on one machine and fails on another.

## What happened

A subject of 54 characters, rejected as 86:

```
$ ./commit-msg msg-subject.txt
❌ Subject is 86 characters (max 72).

feat(mocks): филтровият ред на S3 и празното след него

Shorten the subject; move the detail into the body.
$ echo $?
1
```

The subject is 54 characters. Nothing in the output accounts for the other 32, and the word `characters` actively points away from the answer.

The same string as a body line, on the same machine, differs only by locale — file unchanged between the two runs:

```
$ LC_ALL=C       ./commit-msg msg-body.txt
❌ Commit body lines should wrap at 72 characters:
  line 3 (86 chars)

Wrap the description text. URLs and footers are exempt.
$ echo $?
1

$ LC_ALL=C.UTF-8 ./commit-msg msg-body.txt
✅ Commit message format validated (RuleOfCode v7.19.0)
$ echo $?
0
```

And under that same UTF-8 locale where the body passes, the identical string still fails as a subject:

```
$ LC_ALL=C.UTF-8 ./commit-msg msg-subject.txt
❌ Subject is 86 characters (max 72).
$ echo $?
1
```

So one string, one limit of `72`, three different verdicts depending on where it sits and what `LANG` is set to.

## What is wrong

Two length checks in one hook, each using a different tool, neither one measuring characters reliably.

`dist/hooks/installer.js:224` — the subject:

```sh
SUBJECT_LEN=$(printf '%s' "$FIRST_LINE" | wc -c | tr -d ' ')
if [ "$SUBJECT_LEN" -gt 72 ]; then
    echo "❌ Subject is $SUBJECT_LEN characters (max 72)."
```

`wc -c` counts **bytes** by POSIX definition, in every locale. `wc -m` is the character counter.

`dist/hooks/installer.js:249` — the body:

```sh
LONG_LINES=$(printf '%s\n' "$COMMIT_MSG" | awk 'NR>=3 { if (length($0) > 72 && ... ) print "  line " NR " (" length($0) " chars)" }')
```

`awk`'s `length()` counts characters **when the locale is UTF-8**, and bytes otherwise.

Measured on one file, `feat(mocks): филтровият ред на S3 и празното след него` — 54 characters, 86 bytes:

| `LC_ALL` | `wc -c` | `wc -m` | `awk length()` |
| --- | --- | --- | --- |
| unset (this machine's default) | 86 | 86 | 86 |
| `C` | 86 | 86 | 86 |
| `C.UTF-8` | 86 | **54** | **54** |
| `en_US.UTF-8` | 86 | **54** | **54** |
| `bg_BG.UTF-8` | 86 | **54** | **54** |

Three consequences follow, and they are separable:

1. **The subject limit is a byte limit, always.** At 72, that is 72 characters of ASCII, ~36 of Cyrillic or Greek, ~24 of CJK. The limit silently means something different per team.
2. **The subject message misnames its own unit.** It says `characters` while holding a byte count, so the author cannot reconcile the number with what they see and ends up deleting words until it passes.
3. **The body limit is not stable across machines.** A body line that passes on a developer's UTF-8 shell fails on a CI runner with `LANG` unset — same commit, same hook, same version. `.husky` hooks run under whatever `git` inherits, which nothing here pins.

## Reproduction

From an empty directory; no repository or config needed, since the hook is a standalone script:

```shell
mkdir roc-subject && cd roc-subject
cp /path/to/project/.husky/commit-msg ./hook && chmod +x ./hook
# or generate a fresh one: npx ruleofcode reinstall-hooks --force

printf 'feat(mocks): филтровият ред на S3 и празното след него\n' > msg-subject.txt
{ printf 'feat(mocks): kratko\n\n'
  printf 'feat(mocks): филтровият ред на S3 и празното след него\n'; } > msg-body.txt

./hook msg-subject.txt              ; echo "exit=$?"   # ❌ "86 characters", exit 1
LC_ALL=C       ./hook msg-body.txt  ; echo "exit=$?"   # ❌ "line 3 (86 chars)", exit 1
LC_ALL=C.UTF-8 ./hook msg-body.txt  ; echo "exit=$?"   # ✅ exit 0
LC_ALL=C.UTF-8 ./hook msg-subject.txt; echo "exit=$?"  # ❌ still 86, exit 1
```

Any non-ASCII subject works; the string above is a real rejected commit from this project.

## What it cost me

This repository's commit language is Bulgarian, so every subject is Cyrillic and the real budget is ~36 characters, not 72. That is under half of what the rule announces, and it is enough that a conventional-commit prefix plus a scope leaves room for about four words.

The expensive part is not the limit, it is the mislabelled number. The first several times, I counted 54, read `86`, and assumed I had miscounted — the fix appears to be "delete words until it stops complaining", which is what I did. I now keep a note to myself that the subject limit is **72 bytes**, because the tool will not tell me that. A rule that has to be documented privately against its own error message is not doing the job the message is for.

Worth noting the diagnosis went the wrong way first: with `LANG` unset here, `wc -m` also returns 86, so my initial reading was that the two checks agreed. They only diverge once a UTF-8 locale is present, which is the case on most CI images and developer machines but not in this git-bash. The locale table above is what settled it.

## Is the same logic somewhere else?

Checked. `wc -c` appears once in the generator, at `:224`. `awk length()` appears once, at `:249`. The `pre-commit` and `pre-push` templates contain no length measurement at all:

```
$ grep -n "wc -c\|wc -m" .husky/pre-commit .husky/pre-push
(no output)
```

So this is confined to the one hook, in the two places above.

## Suggested fix

Decide the unit once, then make both checks use it and both messages say it.

If the intent is characters, `awk` already does it correctly under a UTF-8 locale, and reusing it makes the two halves agree by construction:

```diff
-    SUBJECT_LEN=$(printf '%s' "$FIRST_LINE" | wc -c | tr -d ' ')
+    SUBJECT_LEN=$(printf '%s\n' "$FIRST_LINE" | awk '{print length($0)}')
```

`wc -m` also works, with the same caveat as `awk`: it falls back to bytes when the locale is not UTF-8 (86 rather than 54 in the table above). Either way, the locale needs pinning inside the hook so the gate does not change meaning between a laptop and a CI runner:

```sh
export LC_ALL="${LC_ALL:-C.UTF-8}"
```

If the byte limit is deliberate — protecting terminals, older git tooling, or `git log --oneline` alignment — then the subject message should say `bytes`, and printing both numbers would make it actionable rather than puzzling:

```
❌ Subject is 86 bytes / 54 characters (max 72 bytes).
```

That form also removes the need for the reader to know which unit is in play.

## What I did NOT verify

- Whether this reproduces on 7.22.0. This project is pinned to 7.21.0 and I have not staged the fix on a newer release.
- Whether the subject limit being bytes is intentional. I found no statement either way in the hook, and I did not read the CHANGELOG history for the commit that introduced `wc -c`.
- What `LANG` actually is on a typical CI runner for this project. The cross-locale claim is measured on this machine by setting `LC_ALL` explicitly; I have not observed the failure on a real CI run.
- Whether other shells' builtins (`dash`, `busybox awk`) behave as GNU awk does here. Measured with GNU coreutils 8.32 and GNU Awk 5.4.0 only.
- Non-BMP characters (emoji, 4-byte sequences). All measurements above are 2-byte Cyrillic; I did not test whether `awk length()` counts an astral character as one or two.
- The hook I exercised self-reports `v7.19.0`. Under 7.21.0 it is byte-identical apart from the three version strings, which I confirmed with `reinstall-hooks --force` and `diff` — that staleness is a separate defect, #129.

## My config

```json
"hooks": {
  "preCommit": true,
  "prePush": true,
  "commitMsg": true
},
"project": {
  "name": "festrr-mobile",
  "type": "angular"
}
```

Environment: Windows 11, git-bash (`GNU bash 5.3.9`, cygwin), Node v24.16.0, npm 11.13.0, GNU coreutils 8.32, GNU Awk 5.4.0. `LANG` and `LC_ALL` both unset. Nx monorepo, run from a linked git worktree.

**Comment by ruleofcodedev** [comment]
Related delivery problem, filed separately as #129: `validateHooks()` checks `commit-msg` against the literal string `'RuleOfCode'` instead of the version (`installer.js:380`, where the other two hooks use ``v${this.VERSION}``), so that hook is never reported outdated and `reinstall-hooks` leaves it untouched.

Practical consequence for this issue: once the `wc -c` fix ships, existing projects will **not** pick it up — `reinstall-hooks` will keep printing "All git hooks are up to date" while running the old validation. Anyone verifying the fix on an existing checkout needs `reinstall-hooks --force`.

**Comment by ruleofcodedev** [comment]
Fixed in #136 — open, CI running. Merging as soon as it is green.

Both halves: subject and body are measured the same way (continuation bytes deleted, `LC_ALL=C` pinned). `.husky/commit-msg` in this repository is regenerated in the same commit, so our own gate measures what it claims to.

`wc -m` would only have moved the problem — it counts bytes too under a C locale.

**Comment by ruleofcodedev** [comment]
Merged in #136 (`763e21d`), on `develop`. Ships in 7.23.0.

Both measurements: continuation bytes deleted, `LC_ALL=C` pinned on each. `.husky/commit-msg` in this repository is regenerated in the same commit, so our own gate measures what it claims to.

Worth pairing this with #129 when you read the release notes — that one is why this fix would have reached almost nobody. `validateHooks()` checked the installed commit-msg hook against the literal `'RuleOfCode'`, which every generated hook has ever contained, so `reinstall-hooks` reported a hook of any age as up to date and left it in place. Fixed in #137.
